crates/oxc_transformer/src/es2015/arrow_functions.rs

```diff
@@ -324,7 +246,76 @@ impl<'a> ArrowFunctions<'a> {
             ));
         }
         let this_var = this_var.as_ref().unwrap();
-        this_var.create_spanned_read_reference(span, ctx)
+        Some(this_var.create_spanned_read_reference(span, ctx))
+    }
+
+    /// Find arrow function we are currently in, if it's between current node, and where `this` is bound.
+    /// Return its `ScopeId`.
+    fn get_arrow_function_scope(ctx: &mut TraverseCtx<'a>) -> Option<ScopeId> {
+        // `this` inside a class resolves to `this` *outside* the class in:
+        // * `extends` clause
+        // * Computed method key
+        // * Computed property key
+        //
+        // ```js
+        // // All these `this` refer to global `this`
+        // class C extends this {
+        //     [this] = 123;
+        //     static [this] = 123;
+        //     [this]() {}
+        //     static [this]() {}
+        // }
+        // ```
+        //
+        // `this` resolves to the class / class instance (i.e. `this` defined *within* the class) in:
+        // * Class method bodies
+        // * Class property bodies
+        // * Class static blocks
+        //
+        // ```js
+        // // All these `this` refer to `this` defined within the class
+        // class C {
+        //     a = this;
+        //     static b = this;
+        //     #c = this;
+        //     d() { this }
+        //     static e() { this }
+        //     #f() { this }
+        //     static { this }
+        // }
+        // ```
+        //
+        // So in this loop, we only exit when we encounter one of the above.
+        for ancestor in ctx.ancestors() {
+            match ancestor {
+                // Top level
+                Ancestor::ProgramBody(_)
+                // Function (includes class method body)
```

Comment:
Is it possible to implement it by iterating over ancestors scopes?

Reply:
Unfortunately not, in order to pass these 2 tests (which Babel didn't have, but I added):

* [this-in-class-property/input.js](https://github.com/oxc-project/oxc/blob/6e0c12f5c3a552fece9773be95b56245d05389b1/tasks/transform_conformance/tests/babel-plugin-transform-arrow-functions/test/fixtures/this-in-class-property/input.js)
* [this-in-class-property-key/input.js](https://github.com/oxc-project/oxc/blob/6e0c12f5c3a552fece9773be95b56245d05389b1/tasks/transform_conformance/tests/babel-plugin-transform-arrow-functions/test/fixtures/this-in-class-property-key/input.js)

To get both of them right, we have to figure out whether `this` is in:

1. Class property value - `class { x = this; }` or
2. Class property key - `class { [this] = 1; }`

There is no `ScopeFlags::ClassProperty`, so only way to tell the difference is to go up the ancestor chain.

Reply:
It sounds like we can combine these two ways. Let's try it in another PR if we want.